File: day7/day7.py
from pathlib import Path
from statistics import median, mean
import math

path = Path("input.txt")

with open(path, 'r') as f:
    data = list(map(int, f.readline().strip().split(",")))

# Part 1


def fuel_cost(data):
    med = median(data)
    return int(sum(abs(i - med) for i in data))


print(fuel_cost(data))

# Part 2


def cost(dist):
    return int(dist * (dist + 1) / 2)


def sum_cost(arr, val):
    return sum(cost(abs(elem - val)) for elem in arr)


def fuel_cost2(data):
    window = int(len(data) * .1)
    m = round(mean(data))
    minim = sum_cost(data, m)
    for i in range(m - window, m + window):
        if (temp := sum_cost(data, i)) < minim:
            minim = temp
    return minim


print(fuel_cost2(data))

File: day7/day7.jl
using Statistics

data = parse.(Int, split(readline(joinpath(@__DIR__, "input.txt")), ","))

# Part 1

cost1(arr, val) = sum(abs.(arr .- val))
println(cost1(data, convert(Int, median(data))))

# Part 2

gauss_sum(dist) = dist .* (dist .+ 1) ./ 2
cost2(arr, val) = sum(gauss_sum(abs.(arr .- val)))

c1 = cost2(data, round(mean(data) - 1/2))
c2 = cost2(data, round(mean(data) + 1/2))
println(convert(Int, min(c1, c2)))
